packages/viewer/src/components/viewer/viewer.tsx

```diff
@@ -396,77 +422,68 @@ export class Viewer {
       });
   }
 
-  private unload(): void {
-    throw new UnsupportedOperationError('Unsupported operation.');
-  }
-
   private async handleStreamRequest(
     request: vertexvis.protobuf.stream.IStreamRequest
   ): Promise<void> {
     if (request.drawFrame != null) {
-      this.drawFrame(request.drawFrame);
+      this.handleFrame(request.drawFrame);
     } else if (request.gracefulReconnection != null) {
-      await this.reconnectStreamingClient(
-        this.resource,
-        request.gracefulReconnection.streamId.hex
-      );
+      this.handleGracefulReconnect(request.gracefulReconnection);
     }
   }
 
-  private async drawFrame(
-    frame: vertexvis.protobuf.stream.IDrawFramePayload
+  private handleGracefulReconnect(
+    payload: vertexvis.protobuf.stream.IGracefulReconnectionPayload
+  ): void {
+    this.reconnectStreamingClient(this.resource, payload.streamId.hex);
+  }
+
+  private async handleFrame(
+    payload: vertexvis.protobuf.stream.IDrawFramePayload
   ): Promise<void> {
-    const frameNumber = this.lastFrameNumber + 1;
+    const frame = Frame.fromProto(payload);
+    const frameNumber = frame.sequenceNumber;
+    this.frameReceived?.emit(frame);
 
-    const image = await this.loadImageBytes(frame.image);
+    const image = await this.loadImageBytes(payload.image);
 
-    if (frameNumber > this.lastFrameNumber) {
+    if (this.lastFrameNumber == null || frameNumber > this.lastFrameNumber) {
       this.lastFrameNumber = frameNumber;
-      this.frameAttributes = Frame.fromProto(frame);
-
-      this.drawImage(
-        image,
-        frame.imageAttributes.frameDimensions,
-        frame.imageAttributes.imageRect,
-        frame.imageAttributes.scaleFactor
-      );
-
-      this.frameReceived?.emit(this.frameAttributes);
+      this.lastFrame = frame;
+      this.drawImage(image, frame);
     }
 
     image.dispose();
   }
 
-  private drawImage(
-    image: LoadedImage,
-    sceneViewport: vertexvis.protobuf.stream.IDimensions,
-    imagePosition: vertexvis.protobuf.stream.IRectangle,
-    scaleFactor: number
-  ): void {
+  private drawImage(image: LoadedImage, frame: Frame.Frame): void {
     if (this.canvasElement != null) {
       const context = this.canvasElement.getContext('2d');
 
       if (context != null && this.dimensions != null) {
+        const { imageAttributes } = frame;
         const imageRect = vertexvis.protobuf.stream.Rectangle.fromObject(
-          sceneViewport
+          imageAttributes.frameDimensions
         );
         const fitTo = Rectangle.fromDimensions(this.dimensions);
         const fit = Rectangle.containFit(fitTo, imageRect);
 
         const scaleX = fit.width / imageRect.width;
         const scaleY = fit.height / imageRect.height;
 
-        const startXPos = imagePosition.x * scaleX;
-        const startYPos = imagePosition.y * scaleY;
+        const startXPos = imageAttributes.imageRect.x * scaleX;
```

Comment:
Will these changes resolve this issue? https://vertexvis.atlassian.net/browse/SDK-909

Reply:
Not yet. Have no idea what's causing that annoying bug.